server/src/main/java/org/opensearch/indices/IndicesRequestCache.java

```diff
@@ -504,23 +528,41 @@ private void updateCleanupKeyToCountMapOnCacheEviction(CleanupKey cleanupKey) {
             }
             ShardId shardId = indexShard.shardId();
 
-            cleanupKeyToCountMap.computeIfPresent(shardId, (shard, keyCountMap) -> {
-                keyCountMap.computeIfPresent(cleanupKey.readerCacheKeyId, (key, currentValue) -> {
-                    // decrement the stale key count
+            cleanupKeyToCountMap.compute(shardId, (key, readerCacheKeyMap) -> {
+                if (readerCacheKeyMap == null || !readerCacheKeyMap.containsKey(cleanupKey.readerCacheKeyId)) {
+                    // If ShardId is not present or readerCacheKeyId is not present
+                    // it should have already been accounted for and hence been removed from this map
+                    // so decrement staleKeysCount
                     staleKeysCount.decrementAndGet();
-                    int newValue = currentValue - 1;
-                    // Remove the key if the new value is zero by returning null; otherwise, update with the new value.
-                    return newValue == 0 ? null : newValue;
-                });
-                return keyCountMap;
+                    // Returning null removes the entry for the shardId, if it exists
+                    return null;
```

Comment:
whatever we return here will be the final value in the map, since it was already null we are returning null again. 

I clearly see the comment `// Returning null removes the entry for the shardId, if it exists` misleading here

Reply:
That's a great catch, i fixed this and improved tests to go over every scenario of this method. see the latest commit [`581ea2a` (#13070)](https://github.com/opensearch-project/OpenSearch/pull/13070/commits/581ea2afdbb3d3618e31fd517c06ff2880bf6ff7)